Std/Logic.lean

```diff
@@ -732,6 +732,79 @@ theorem proof_irrel_heq {p q : Prop} (hp : p) (hq : q) : HEq hp hq := by
 theorem congrArg₂ (f : α → β → γ) {x x' : α} {y y' : β}
     (hx : x = x') (hy : y = y') : f x y = f x' y' := by subst hx hy; rfl
 
+theorem congrFun₂ {β : α → Sort _} {γ : ∀ a, β a → Sort _}
+    {f g : ∀ a b, γ a b} (h : f = g) (a : α) (b : β a) :
+    f a b = g a b :=
+  congrFun (congrFun h _) _
+
+theorem congrFun₃ {β : α → Sort _} {γ : ∀ a, β a → Sort _} {δ : ∀ a b, γ a b → Sort _}
+      {f g : ∀ a b c, δ a b c} (h : f = g) (a : α) (b : β a) (c : γ a b) :
+    f a b c = g a b c :=
+  congrFun₂ (congrFun h _) _ _
+
+theorem funext₂ {β : α → Sort _} {γ : ∀ a, β a → Sort _}
+    {f g : ∀ a b, γ a b} (h : ∀ a b, f a b = g a b) : f = g :=
+  funext fun _ => funext <| h _
+
+theorem funext₃ {β : α → Sort _} {γ : ∀ a, β a → Sort _} {δ : ∀ a b, γ a b → Sort _}
+    {f g : ∀ a b c, δ a b c} (h : ∀ a b c, f a b c = g a b c) : f = g :=
+  funext fun _ => funext₂ <| h _
+
+theorem ne_of_apply_ne {α β : Sort _} (f : α → β) {x y : α} : f x ≠ f y → x ≠ y :=
+  mt <| congrArg _
+
+protected theorem Eq.congr (h₁ : x₁ = y₁) (h₂ : x₂ = y₂) : x₁ = x₂ ↔ y₁ = y₂ := by
+  subst h₁; subst h₂; rfl
+
+theorem Eq.congr_left {x y z : α} (h : x = y) : x = z ↔ y = z := by rw [h]
+
+theorem Eq.congr_right {x y z : α} (h : x = y) : z = x ↔ z = y := by rw [h]
+
+alias congr_arg := congrArg
+alias congr_arg₂ := congrArg₂
+alias congr_fun := congrFun
+alias congr_fun₂ := congrFun₂
+alias congr_fun₃ := congrFun₃
+
+theorem eq_mp_eq_cast (h : α = β) : Eq.mp h = cast h :=
+  rfl
+
+theorem eq_mpr_eq_cast (h : α = β) : Eq.mpr h = cast h.symm :=
+  rfl
```

Comment:
I'm not thilled about PRs that upstream lemmas but drop `@[simp]` attributes.

Reply:
I think that was an oversight.  We could add them in a separate PR.

Reply:
This was my bad! The issue is that Std doesn't know much what to do with `cast`, so these simps led "nowhere" in a sense.